src/tracer.cc

```diff
@@ -369,30 +371,225 @@ TraceManager::Trace::InitTracer(
+      true /*is_root_span*/);
+  // Initializing OTel context and storring "InferRequest" span as a root span
+  // to keep it alive for the duration of the request.
+  otel_context_ = opentelemetry::context::Context({kRootSpan, root_span});
 }
 
-void
-TraceManager::Trace::InitSpan(const otel_common::SystemTimestamp& timestamp_ns)
+opentelemetry::nostd::shared_ptr<otel_trace_api::Span>
+TraceManager::Trace::InitSpan(
+    std::string name, const otel_common::SystemTimestamp& timestamp_ns,
+    const uint64_t& raw_timestamp_ns, bool is_root_span,
+    std::string parent_span_key)
 {
+  opentelemetry::nostd::shared_ptr<otel_trace_api::Span> span{nullptr};
   otel_trace_api::StartSpanOptions options;
-  options.kind = otel_trace_api::SpanKind::kServer;  // server
+  options.kind = otel_trace_api::SpanKind::kServer;
   options.start_system_time = timestamp_ns;
-  // [FIXME] think about names
-  trace_span_ =
-      provider_->GetTracer("triton-server")->StartSpan("InferRequest", options);
+  options.start_steady_time =
+      otel_common::SteadyTimestamp{std::chrono::nanoseconds{raw_timestamp_ns}};
+
+  // If the new span is a child span, we need to retrieve its parent from
+  // the context and provide it through StartSpanOptions to the child span
+  if (!is_root_span && otel_context_.HasKey(parent_span_key)) {
+    auto parent_span = opentelemetry::nostd::get<
+        opentelemetry::nostd::shared_ptr<otel_trace_api::Span>>(
+        otel_context_.GetValue(parent_span_key));
+    options.parent = parent_span->GetContext();
+  }
+  span = provider_->GetTracer(kTritonTracer)->StartSpan(name, options);
+  return span;
+}
+
+void
+TraceManager::Trace::EndSpanNow(
+    opentelemetry::nostd::shared_ptr<otel_trace_api::Span> span)
+{
+  if (span != nullptr) {
+    auto timestamp_ns = std::chrono::duration_cast<std::chrono::nanoseconds>(
+                            std::chrono::steady_clock::now().time_since_epoch())
+                            .count();
+    EndSpan(span, timestamp_ns);
+  }
+}
+
+
+void
+TraceManager::Trace::EndSpan(
+    opentelemetry::nostd::shared_ptr<otel_trace_api::Span> span,
+    const uint64_t& raw_timestamp_ns)
+{
+  if (span != nullptr) {
+    otel_trace_api::EndSpanOptions end_options;
+    end_options.end_steady_time = otel_common::SteadyTimestamp{
+        std::chrono::nanoseconds{raw_timestamp_ns}};
+    span->End(end_options);
+  }
 }
 
 void
-TraceManager::Trace::EndSpan()
+TraceManager::Trace::ReportToOpenTelemetry(
+    TRITONSERVER_InferenceTrace* trace,
+    TRITONSERVER_InferenceTraceActivity activity, uint64_t timestamp_ns)
+{
+  auto current_span_key = GetSpanNameForActivity(activity);
+  if (current_span_key.empty()) {
+    return;
+  }
+  uint64_t id;
+  LOG_TRITONSERVER_ERROR(
+      TRITONSERVER_InferenceTraceId(trace, &id), "getting trace id");
+  current_span_key = current_span_key + std::to_string(id);
+  switch (activity) {
+    case TRITONSERVER_TRACE_REQUEST_START: {
+      const char* model_name;
```

Comment:
Create new function for the case blocks. 
It is a good practice to have shorter case blocks. Makes the code more readable.

Reply:
I've decided to get rid of switch-case altogether